### PR TITLE
Remove generated files from source control

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Remove automatically generated files from source control.

The following files are no longer checked into git, and must instead be generated automatically before building the library or its companion programs:
* `library/error.c`
* `programs/test/query_config.c`
* `library/version_features.c`
* `visualc/VS2010/*`
* `programs/psa/psa_constant_names_generated.c`
* some `tests/suites/test_suite_psa_crypto_*.*.data`

This comprises the following subtasks:
* [x] Remove the generated files.
* [x] Makefile targets for the generated files.
* [x] With make, `make neat` runs `make clean` and also removes configuration-and-platform-independent generated files.
* [x] Generate the generated files before running any component in `all.sh`.
* [x] Generate the generated files before doing a Windows build on Travis.
* [x] Generate the generated files before doing a Windows build on Jenkins. This needs both work in this PR so that generating the files works on Windows, and adding a step to the Jenkins Groovy scripts (done in [private link](https://github.com/ARMmbed/mbedtls-test/pull/177)).
* [x] A mechanism to mitigate the risk of regressions in generated files.
    * [x] For storage format test cases: https://github.com/ARMmbed/mbedtls/pull/4408
    * [x] For everything else I don't think there is any particular risk of regression that would not break the build or be easily noticeable in review.
* [x] Documentation.

Status: ready.

Follow-ups, out of scope of this pull request:
* CMake targets for the generated files. → https://github.com/ARMmbed/mbedtls/issues/4474
* With CMake, `make neat` runs `make clean` and also removes configuration-and-platform-independent generated files. → https://github.com/ARMmbed/mbedtls/issues/4474
* **Releases will include the generated files** so that users don't need Perl or Python to build the library. The tooling for that is out of scope here. → https://github.com/ARMmbed/mbedtls/issues/4475
* Better support for cross-compilation. → https://github.com/ARMmbed/mbedtls/issues/4528
* Better testing of the various ways to generate files. → https://github.com/ARMmbed/mbedtls/issues/4535

### CODE REVIEW

File: Makefile
Comment (mpg) [inline Makefile:47]: Note: do we really need a Gnu Makefile target for generating Visual C files? What problem does it solve, who's going to use it? I don't have problem with having one, except apparently we can't easily have proper dependencies, which are most of the point of `make` as opposed to a script, so why not skip this target and just let people invoke the script directly?

File: library/Makefile
Comment (mpg) [inline library/Makefile:290]: OTOH, when switching configurations, you probably want to re-build most if not all of the C files anyway, so re-generating this one is IMO lost in the noise... but since we don't have proper dependencies in the Makefiles in general anyway, I don't object to omitting this one too.

**Review by gilles-peskine-arm — COMMENTED** [review]
I hereby approve Bence's changes.

**Comment by bensze01** [comment]
@gilles-peskine-arm Hey, Gilles I've added two more commits you should look at.

**Comment by bensze01** [comment]
Here is a successful CI run with the latest changes: https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-ci-testing/252/

File: scripts/mbedtls_dev/c_build_helper.py
Comment (gilles-peskine-arm) [inline scripts/mbedtls_dev/c_build_helper.py:159]: The `.obj` file should be removed in the `finally` cleanup, not just on the success path.

File: scripts/mbedtls_dev/c_build_helper.py
Comment (gilles-peskine-arm) [inline scripts/mbedtls_dev/c_build_helper.py:136]: Caching `_cc_is_msvc` is fragile: maybe at some point we'll use this function in a script that does builds with multiple compilers. I don't plan to do this anywhere soon, we can live with caching, but please add a note to this function's documentation.

File: scripts/mbedtls_dev/c_build_helper.py
Comment (gilles-peskine-arm) [inline scripts/mbedtls_dev/c_build_helper.py:149]: Is there no way to avoid the generation of the `.obj` file, e.g. `-FoNUL`?

File: scripts/mbedtls_dev/c_build_helper.py
Comment (gilles-peskine-arm) [inline scripts/mbedtls_dev/c_build_helper.py:52]: How about not bothering with `obj_name` here and using `exe_name[:-3]obj` in the caller?

File: scripts/mbedtls_dev/c_build_helper.py
Comment (bensze01) [reply scripts/mbedtls_dev/c_build_helper.py:149]: No, sadly `cl.exe` only produces the `.obj` file, and calls l`ink.exe` to actually create the executable.
If we specify `-FoNUL`, the link step will fail.

File: scripts/mbedtls_dev/c_build_helper.py
Comment (gilles-peskine-arm) [reply scripts/mbedtls_dev/c_build_helper.py:136]: Why bother with caching anyway? I've been treating the CC calls as an expensive operation. `test_psa_constant_names.py` only calls this function once. `generate_psa_tests.py` currently calls this function multiple times, but that's only because of a bug in its own caching mechanism which #4385 is fixing. So I think we shouldn't bother with caching here at all.

**Comment by bensze01** [comment]
@gilles-peskine-arm I think I've addressed all your feedback.

File: scripts/mbedtls_dev/c_build_helper.py
Comment (bensze01) [reply scripts/mbedtls_dev/c_build_helper.py:138]: Ok, in that case, I'll remove the caching.

**Comment by bensze01** [comment]
Ok, I think I've addressed all the feedback.

**Review by gilles-peskine-arm — COMMENTED** [review]
Thanks for the updates, looks good to me. I've kick off a [CI job with the updated Windows test scripts](https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-ci-testing/254/).

**Comment by bensze01** [comment]
@gilles-peskine-arm Ah, sorry, the CI run you started referred to the wrong Mbed TLS repo (mine), where I didn't upload the latest changes at the time.

I've aborted it, and kicked off a [CI job with the correct branches](https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-ci-testing/255/parameters/)

**Comment by bensze01** [comment]
The CI run linked in the comment above completed successfully.

**Comment by mpg** [comment]
Note: this still targets `development_3.0`, will need to change before merging. (Also, resolve conflicts with the new target.)

**Review by bensze01 — APPROVED** [review]
I have reviewed Gilles's commits, and they look good to me.

The PR still needs to be re-targeted to `development` before it can be merged, as Manuel indicated above.

**Comment by gilles-peskine-arm** [comment]
I've pushed a rebased version to resolve merge conflicts, and I changed the base branch to `development`. The previous history is in https://github.com/gilles-peskine-arm/mbedtls/tree/no-generated-files-3.0-3. The conflicts in the rebase were changes in `programs/Makefile` that interfered with the removal of `aescrypt2`, and delete/modify conflicts for generated files that had changed on `development`.

The next step is to add missing documentation. I intend no other changes except as mandated by review feedback (or CI feedback, but I expect CI to pass now).

**Comment by bensze01** [comment]
The DCO check doesn't seem to get re-run when you change the base branch of the patch - it failed due to a couple of DCO-less commits we accidentally let into development when merging #4397.

We could try to resolve this by pushing an empty dummy commit on top of this branch, and force-pushing the original HEAD afterwards to remove the dummy commit.

**Comment by gilles-peskine-arm** [comment]
I've pushed documentation commits and DCO is now passing.

File: programs/Makefile
Comment (mpg) [inline programs/Makefile:58]: Missing word I think: "still works _if_ you tweak".

File: README.md
Comment (mpg) [inline README.md:73]: I don't think that works without [the first commit of 4514](https://github.com/ARMmbed/mbedtls/pull/4514/commits/7f14c433513409d993f4a1dcf6d6912a0fcd9707) - which you might want to cherry-pick here for consistency with the documentation.

**Review by mpg — DISMISSED** [review]
Looks good to me except for some minor details.

File: ChangeLog.d/no-generated-files.txt
Comment (mpg) [inline ChangeLog.d/no-generated-files.txt:6]: From the CI: `assemble_changelog.py` doesn't like lines that are longer than 80 bytes, and doesn't seem to care that those 81 bytes are actually 79 chars.

File: README.md
Comment (gilles-peskine-arm) [reply README.md:73]: Good find, but I don't like the way you do it: it causes `check-generated-files -u` to always print messages about the files being out of date, even when there are no changes. So I did it slightly differently.

I didn't fix the annoyance that `check-generated-files -u` always updates the mtime on the files, even if they end up with the same content. I consider this out of scope, and less important now that you can use `make` if you care about timestamps.

File: ChangeLog.d/no-generated-files.txt
Comment (gilles-peskine-arm) [reply ChangeLog.d/no-generated-files.txt:6]: That annoyed me and I went and made the script work on UTF-8 instead of bytes.

**Review by bensze01 — CHANGES_REQUESTED** [review]
The following commits at the head of the branch seem to have been left out of the rebase:

19:  b2771c0b6 <  -:  --------- Detect MSVC without relying on compiler filename
20:  49b4366d9 <  -:  --------- Clean up object files produced by MSVC
21:  39f05c426 <  -:  --------- Add notice of caching whether the compiler is MSVC
22:  9f258f118 <  -:  --------- Move object file handling out of create_c_file()
23:  df2619edc <  -:  --------- Remove object file in finally
24:  182831545 <  -:  --------- Remove caching of cc_is_msvc

Also, please remove `tests/git-scripts/pre-commit.sh` as well, since we no longer need to check the generated files before commiting.

**Comment by gilles-peskine-arm** [comment]
Sorry about the missing commits: I'd been working from a stale local copy. To minimize the number of force-pushes to the pull request, I've cherry-picked them to immediately after my last push.

I also fixed a few issues I noticed, including one which led me to revert an earlier commit. We may want to remove "Create visualc directory if it doesn't exist" and its reversal before merging, but I'm refraining from doing it now to facilitate review.

File: Makefile
Comment (gilles-peskine-arm) [inline Makefile:132]: `make neat` is close to what [GNU](https://www.gnu.org/software/make/manual/html_node/Standard-Targets.html#Standard-Targets) calls `maintainer-clean`, but not exactly the same. It may be argued to be the same as `distclean` instead. Should I use my “custom” name `neat` to avoid confusion, or a more standard name to improve discoverability?

File: Makefile
Comment (bensze01) [inline Makefile:48]: If we only care about the fact these files exist, not their contents, wouldn't an order-only dependency suffice?
```suggestion
$(VISUALC_FILES): | library/error.c library/version_features.c
```

**Review by bensze01 — CHANGES_REQUESTED** [review]

**Comment by gilles-peskine-arm** [comment]
There are remove/modify conflicts in the generated files, and this is likely to continue until this PR is merged. Rather than keep rebasing again and again, I propose to do a merge once this PR has been approved. I've pushed the current version of this merge to https://github.com/ARMmbed/mbedtls/tree/dev/gilles-peskine-arm/no-generated-files-3.0 and a [Travis job](https://travis-ci.com/github/ARMmbed/mbedtls/builds/226306967) is now running.

**Review by bensze01 — APPROVED** [review]
Thanks for the changes Gilles, Looks good to me.

File: Makefile
Comment (mpg) [reply Makefile:131]: I like `neat`. If we're not sure the meaning is the same as `maintainer-clean` or `distclean` then it's probably better to have a distinct name. I don't think discoverability matters that much considering it's only for people using the development branch as opposed to releases.

File: scripts/mbedtls_dev/c_build_helper.py
Comment (mpg) [reply scripts/mbedtls_dev/c_build_helper.py]: Note: if we were doing caching, I think I'd be more inclined to use the facilities of [functools](https://docs.python.org/3/library/functools.html) than rolling our own (with the associated pylint warnings to silence).

**Review by mpg — APPROVED** [review]
I've reviewed the rebase as well as the additional commits and it all looks good to me.

**Comment by mpg** [comment]
Regarding the merge strategy, should we request a short merge freeze on development while the CI runs on the merge of the approved version of this PR with the latest development?

**Comment by gilles-peskine-arm** [comment]
I have restructured the commits as follows:
* Moved “Ignore generated source files that are no longer checked in” and “Move VS2010 Git ignore list to the VS2010 directory” earlier, allowing “Create visualc directory if it doesn't exist” and its reversal to be removed.
* Squashed “Typo in comment” into the commit that introduced the typo (“Don't require make to generate visualc files”).

The end result a13deaf98610c1d3c24549fb9d30476171501143 is identical to the previous final state 502d889de77eeafc2f994df54264eded37594459 (available as https://github.com/gilles-peskine-arm/mbedtls/tree/no-generated-files-3.0-6).

I then merged the latest `development` to resolve the remove/modify conflicts in generated files.

**Review by mpg — APPROVED** [review]
Looks good to me. Regarding the changes since my previous approval, I've checked that:

- `git diff a13deaf 502d889` has an empty output
- the merge commit has correct parents
- the conflict resolutions are correct

**Review by bensze01 — APPROVED** [review]
Looks good to me too:
- The output of `git range-diff development no-generated-files-3.0-6 HEAD^1` aligns with Gilles's comment.
- The output of `diff -u <(git diff $(git merge-base HEAD^{1,2})  HEAD^1) <(git diff HEAD^2 HEAD)` looks correct considering the PRs that were merged since the last rebase.

**Pull request merged**
